extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,163 @@
+ * with the default.
+ * @return {!boolean}
+ */
+function validateRtcConfig(element) {
+  const defaultTimeoutMillis = 1000;
+  const rtcConfig = tryParseJson(
+      element.getAttribute('prerequest-callouts'));
+  if (!rtcConfig) {
+    return false;
+  }
+  try {
+    user().assert(rtcConfig['vendors'] || rtcConfig['urls'],
+                  'RTC Config must specify vendors or urls');
+    user().assert(!rtcConfig['vendors'] || isObject(rtcConfig['vendors']),
+                  'RTC invalid vendors');
+    user().assert(!rtcConfig['urls'] || isArray(rtcConfig['urls']),
+                  'RTC invalid urls');
+  } catch (err) {
+    return false;
+  }
+  let timeout = rtcConfig['timeoutMillis'];
+  if (timeout) {
+    if (!Number.isInteger(timeout) || timeout >= defaultTimeoutMillis || timeout < 0) {
+      timeout = defaultTimeoutMillis;
+      user().warn(TAG, `Invalid RTC timeout: ${timeout}ms, ` +
+                  `using default timeout ${timeoutMillis}ms`);
+    }
+  }
+  rtcConfig['timeoutMillis'] = timeout;
+  return rtcConfig;
+}
+
+function maybeInflatePublisherUrls(rtcConfig, urlToCalloutMap, customMacros, ampDoc) {
+  if (!rtcConfig['urls']) {
+    return;
+  }
+  let url;
+  let remaining;
+  for (let i in rtcConfig['urls']) {
+    url = rtcConfig['urls'][i];
+    maybeInflateAndAddUrl(url, customMacros, ampDoc, urlToCalloutMap);
+    if (Object.keys(urlToCalloutMap).length == MAX_RTC_CALLOUTS) {
+      logRemaining(rtcConfig['urls'].slice(++i), rtcConfig['vendors'])
+      break;
+    }
+  }
+}
+
+function maybeInflateVendorUrls(rtcConfig, urlToCalloutMap, ampDoc) {
+  if (!rtcConfig['vendors'] || Object.keys(urlToCalloutMap).length == MAX_RTC_CALLOUTS) {
+    return;
+  }
+  let vendor;
+  let macros;
+  const vendors = Object.keys(rtcConfig['vendors']);
+  for (let i in vendors) {
+    vendor = vendors[i];
+    macros = rtcConfig['vendors'][vendor]
+    url = RTC_VENDORS[vendor.toLowerCase()];
+    if (url) {
+      maybeInflateAndAddUrl(url, macros, ampDoc, urlToCalloutMap, vendor);
+    }
+    if (Object.keys(urlToCalloutMap).length == MAX_RTC_CALLOUTS) {
+      log_remaining(vendors.slice(i));
+      break;
+    }
+  }
+}
+
+function logRemaining(remaining, opt_vendors) {
+  if (opt_vendors) {
+    remaining = remaining.concat(Object.keys(opt_vendors));
+  }
+  if (remaining) {
+    remaining = JSON.stringify(remaining);
+    dev().warn(TAG, `${MAX_RTC_CALLOUTS} RTC Callout URLS exceeded, ` +
+               ` dropping ${remaining}`);
+  }
+}
+
+/**
+ * Substitutes macros into url, and adds the resulting URL to the list
+ * of callouts. Checks each URL to see if secure. If a supplied macro
+ * does not exist in the url, it is silently ignored.
+ */
+function maybeInflateAndAddUrl(url, macros, ampDoc, urlToCalloutMap, opt_vendor) {
+  const urlReplacements = Services.urlReplacementsForDoc(ampDoc);
+  // TODO: change to use whitelist.
+  url = urlReplacements.expandSync(url, macros);
+  const callout = opt_vendor || url;
```

Comment:
Nit: move this after try/catch given its not needed until then:
urlCalloutMap[url] = opt_vendor || url;